replication-api-impl/src/main/java/com/connexta/replication/api/impl/worker/Worker.java

```diff
@@ -0,0 +1,405 @@
+        try {
+          task.unlock();
+        } catch (InterruptedException ignore) { // fallthrough
+        }
+      }
+      running = false;
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * Set this workers running flag to false if it is not currently transferring an item
+   *
+   * @return {@code true} if the worker was cancelled, otherwise {@code false}
+   */
+  boolean cancelIfNotProcessing() {
+    if (!processing) {
+      this.running = false;
+      return true;
+    }
+    return false;
+  }
+
+  private boolean supported(Task task) {
+    if (OperationType.UNKNOWN.equals(task.getOperation())) {
+      return false;
+    }
+
+    List<MetadataInfo> infos = task.metadatas().collect(Collectors.toList());
+    if (infos.size() > 1) {
+      LOGGER.warn("More than one metadata is currently not supported by this worker");
+      return false;
+    }
+
+    DdfMetadataInfo ddfInfo =
+        infos.stream()
+            .filter(DdfMetadataInfo.class::isInstance)
+            .map(DdfMetadataInfo.class::cast)
+            .findFirst()
+            .orElse(null);
+
+    if (ddfInfo == null) {
+      LOGGER.info("Unable to handle any of the metadatas received in this task");
+      return false;
+    }
+
+    return true;
+  }
+
+  private void handle(Task task) throws InterruptedException {
+    NodeAdapter remote = getDestinationAdapter(task);
+    if (remote == null) {
+      task.unlock();
+      return;
+    }
+
+    try {
+      if (!adaptersAvailable(task, remote)) {
+        return;
+      }
+
+      List<DdfMetadataInfo> infos =
+          task.metadatas()
+              .filter(DdfMetadataInfo.class::isInstance)
+              .map(DdfMetadataInfo.class::cast)
+              .collect(Collectors.toList());
+
+      final DdfMetadataInfo info = infos.get(0);
+      final Metadata metadata =
+          new MetadataImpl(
+              info.getData(), info.getDataClass(), task.getId(), Date.from(info.getLastModified()));
+
+      final OperationType type = task.getOperation();
+      switch (type) {
+        case HARVEST:
+        case REPLICATE_FROM:
+          doCreate(task, metadata, remote, local);
+          break;
+        case REPLICATE_TO:
+          doCreate(task, metadata, local, remote);
+          break;
+        case DELETE:
+          doDelete(task, metadata, remote);
+          break;
+        default:
+          // should never hit this since deserialization will return UNKNOWN
+          task.unlock();
+      }
+    } finally {
+      try {
+        remote.close();
+      } catch (IOException e) {
+        LOGGER.debug("Failed to close destination adapter", e);
+      }
+    }
+  }
+
+  @Nullable
+  private NodeAdapter getDestinationAdapter(Task task) {
+    Site site;
+    final String siteId = task.getQueue().getSite();
+    try {
+      site = siteManager.get(siteId);
+    } catch (NotFoundException e) {
+      LOGGER.debug("Unable to find site {} for queue", siteId, e);
```

Comment:
❓ Should we be failing the task in such a way that it won't be retried if the site no longer exists? Or again completed as a no-op?

Reply:
I don't think you want the broker to have knowledge of the queues it maintains and what to do in such case with the remaining tasks. The broker should just be there to provide access to the queues and let whoever puts stuff in or pulls stuff from them decide. Having to many cooks in the kitchen just complicates everything.